### ISSUE TITLE
Provide guide for installing lobster

### ISSUE THREAD

**Issue by m0d00d** [aardappel/lobster#28]
Currently, it is not obvious how to install lobster on you machine and successfully compile for example a simple hello world program.

**Comment by xyproto** [comment]
After building lobster and finding the `lobster/lobster` executable, I had to be in the same directory as the executable for it to find `stdtype.lobster`, and for this simple "hello world" program to work:

**hello.lobster**:

```py
print("Hello, World!")
```

**Comment by xyproto** [comment]
The include path is hard-coded here: https://github.com/aardappel/lobster/blob/master/dev/src/lobster/lex.h#L54

**Comment by aardappel** [comment]
Yes, I need a better "getting started" section.

@m0d00d at the moment, best is to build the Lobster exe yourself, using instructions on https://htmlpreview.github.io/?https://raw.githubusercontent.com/aardappel/lobster/master/lobster/docs/implementation.html
Though you can also download a pre-build release from github, though those tend to be older.

@xyproto If you look here under "default directories": https://htmlpreview.github.io/?https://raw.githubusercontent.com/aardappel/lobster/master/lobster/docs/command_line_usage.html

You don't need to be in the same dir as the exe: Lobster will load files either relative to the exe, or relative to your .lobster file. In both cases it will also look in an `include` dir under either of those dirs.

Not sure why it didn't work for you.. what platform are you on? what was your command line?

**Comment by xyproto** [comment]
I am on Arch Linux and want to package it, but then I guess Lobster needs to search something like /usr/share/lobster if the local files are missing.

**Comment by aardappel** [comment]
I just tried on both Windows and (Ubuntu) Linux, and I can run it from outside of its dir just fine, e.g. from the parent dir:

```
lobster/lobster lobster/samples/pythtree.lobster
```

That finds `lobster/include` just fine.

No experience with packaging things on Linux, but you'd want to keep the exe and the data together. For comparison, another app of mine https://github.com/aardappel/treesheets is packaged for quite a few distros, and that also accesses data relative to the exe location.

**Comment by m0d00d** [comment]
@aardappel Thanks for the hint. I could manage to compile and run a simple hello world example. 
I think having an easy option like `brew install lobster` could really help in quick starting for OS X.

**Comment by aardappel** [comment]
I agree, that would be nice. But it is early days, so do not want to get bogged down with adminning packages for various platforms at this point.

I will improve the docs for getting started though.

**Comment by xyproto** [comment]
@aardappel Keeping the executable and data together is _possible_ when packaging for Linux, but it's completely against the philosophy of how the file system is laid out. Contrary to on Windows, the clean way to package applications on Linux is to keep the executables in a directory for binaries, like `/usr/bin`, then let the executable search a path like `/usr/share/lobster` for data files, `/etc/lobster` for system configuration and `~/.config/lobster` for per-user configuration. Something along these lines is very common on UNIX-like operating systems. The packages that end up with both executables and data in a directory in `/opt` or `/usr/share` do so only because of necessity, not because it's fits well into what at least I believe to be a pretty and well organized system for installing applications onto a system.

I am convinced that making lobster easy to package on Linux and BSD, perhaps by providing a working `make` and `make install` command, or by using `cmake` or `mesos`, would increase the likelihood of it being picked up by Linux distros.

I do understand that it takes some effort to make Lobster suitable to package for all sorts of systems, though. Are you open for pull requests for making it easier to package in a clean way on at least Arch Linux?

**Comment by aardappel** [comment]
Ok, I can just add `/usr/share/lobster` as one of the paths to find its data on Linux.

CMake likely already generates a `make install` rule, but maybe it needs to be told what files should go where? Basically everything under the `lobster` dir should go into `share` I guess. PRs for this welcome too.

PRs for packaging welcome, especially if it works with multiple distros.

**Comment by aardappel** [comment]
Just added this (untested!) https://github.com/aardappel/lobster/commit/7ee20b185408f0c31c48b42190db33f979b3b471

**Comment by jacereda** [comment]
> Just added this (untested!) [7ee20b1](https://github.com/aardappel/lobster/commit/7ee20b185408f0c31c48b42190db33f979b3b471)

Could that be `#ifndef _WIN32` instead?

**Comment by aardappel** [comment]
Why? I don't think we want to load from `/usr/share/lobster` on Android, iOS, MacOS, Web/Emscripten, etc?

**Comment by jacereda** [comment]
Fair, I'd then try to get rid of the `__linux__` check and always push a value provided at configure time. 

I guess the right thing would be to use here the contents of the CMAKE_INSTALL_DATADIR or CMAKE_INSTALL_FULL_DATADIR to set that location. I'm not very familiar with cmake, but I guess the CMakeLists.txt  should set some symbol in the preprocessor with the contents of one of those in order to use it from the .cpp file. Or maybe I'm just wrong and those defines go straight to the compiler command line.

**Comment by aardappel** [comment]
We can move `/usr/share/lobster` to CMake, but again, only for Linux (or non-Mac other unices). Because most OSes work differently from Linux, we'd still need to use `CMAKE_INSTALL_DATADIR ` or whatever conditional on it being Linux inside the code, or simpler, only apply it if not empty / defined.

Your description and what I found online made it seem like `/usr/share` is pretty universal on these platforms, so it didn't seem that weird to hard-code it.

Either way, the above commit was meant as a quick way to enable packaging on Linux, it can be generalized at any moment that it is actually needed :)

**Comment by jacereda** [comment]
You also want that to be present on Mac, I'm packaging on Mac and I certainly need it. Had to resort to some ugly patching:

https://github.com/NixOS/nixpkgs/pull/59128

**Comment by jacereda** [comment]
`/usr/share` isn't universal at all. I've seen `/opt/share`, `/usr/loca/share`, `/sw/share`... And the weirdest of all, on Nix this would end up on a path like `/nix/store/chc4v581p6q3286m41bh6fc38kbxr4ml-lobster-git/share/`

**Comment by aardappel** [comment]
On the Mac Lobster also builds as a `.app` for when distributing something built with Lobster. That requires the data files to be inside the `.app`. For developer use, is it common for things to be installed in `/usr/share` ? There are developer tools that seem to keep their data inside the `.app` (e.g. CMake).

Ok, so feel free to pull this into CMake in whatever way works best.

**Comment by m0d00d** [comment]
Fixed in [5292e1f](https://github.com/aardappel/lobster/commit/5292e1fced704f2ed5a9aa6bb79bcbc3b2be8888) 👍
